### ISSUE TITLE
Wrong query generation on SubQuery join when having complex key (multiple primary keys)

### ISSUE THREAD

**Issue by guyzilb** [cube-js/cube#6685]
**Describe the bug**

When creating using sub_query while having more than one primary key - the SQL generated for the join is incorrect.

-details-
I have created two separate cubes. 
Our goal is to calculate measures in each cube individually. 
For example, we want to calculate the sum of payments per day for each user and the maximum level per day for each user. 
After calculating those measures in each cube, we would like to use the "sub_query" feature and use a measure from the " level" cube into payment .
This generates SQL code where the join for the subquery is performed based on the primary key fields, which are multiple keys in our case (we have two primary keys for any cube (user_id,event_dt)).
we found the bug inside the generated SQL join for the subquery.

**Create a Cubes:**

```yaml
cubes:
  - name: fact_level_up_pos
    sql:   select * from(
            select  event_dt,new_level_id,user_id , cast(event_ts as date) as event_ss, platform_id_psk
            from public.guy_cube_test_level
            where event_dt >= current_date -10
            )o
    data_source: datasource2

    measures:

      - name: maxLevel
        sql: new_level_id
        type: max

    dimensions:

      - name: event_dt
        sql: event_dt
        type: time
        primary_key: true
        shown: true

      - name: event_ss
        sql: event_ss
        type: time

      - name: user_id
        sql: user_id
        type: string
        primary_key: true
        shown: true

      - name: new_level_id
        sql: new_level_id
        type: number


      - name: platform_id_psk
        sql: platform_id_psk
        type: number





  - name: fact_payment_pos
    sql: select * from public.guy_cube_test_payments
    data_source: datasource2

    joins:
      - name: fact_level_up_pos
        sql: "{CUBE}.user_id = {fact_level_up_pos}.user_id  and {CUBE}.event_dt = {fact_level_up_pos}.event_dt"
        relationship: many_to_one

    measures:

      - name: sum_g
        sql: m_net_amount
        type: sum

      - name: level_max
        sql: "{Level}"
        type: max

    dimensions:

      - name: Level
        sql: "{fact_level_up_pos.maxLevel}"
        type: number
        sub_query: true

      - name: user_id
        sql: user_id
        type: number
        primary_key: true
        shown: true

      - name: app_id_psk
        sql: app_id_psk
        type: number


      - name: app_session_id
        sql: app_session_id
        type: string

      - name: user_id_psk
        sql: user_id_psk
        type: number



      - name: m_net_amount
        sql: m_net_amount
        type: number

      - name: platform_id_psk
        sql: platform_id_psk
        type: number


      - name: event_ts
        sql: event_ts
        type: time


      - name: event_dt
        sql: event_dt
        type: time
        primary_key: true
        shown: true

```
**Bug example:** 
SQL query generated from cube playground
```diff
SELECT
  "fact_payment_pos".user_id "fact_payment_pos__user_id",
  date_trunc(
    'day',
    (
      "fact_payment_pos".event_ts :: timestamptz AT TIME ZONE 'UTC'
    )
  ) "fact_payment_pos__event_ts_day",
  sum("fact_payment_pos".m_net_amount) "fact_payment_pos__sum_g",
  max("fact_payment_pos___level") "fact_payment_pos__level_max"
FROM
  public.guy_cube_test_payments AS "fact_payment_pos"
  LEFT JOIN (
    SELECT
      q_0."fact_payment_pos__user_id",
      q_0."fact_payment_pos__event_dt",
      "fact_level_up_pos__max_level" "fact_payment_pos___level"
    FROM
      (
        SELECT
          "keys"."fact_payment_pos__user_id",
          "keys"."fact_payment_pos__event_dt",
          max(
            "fact_level_up_pos_key__fact_level_up_pos".new_level_id
          ) "fact_level_up_pos__max_level"
        FROM
          (
            SELECT
              DISTINCT "fact_level_up_pos_key__fact_payment_pos".user_id "fact_payment_pos__user_id",
              "fact_level_up_pos_key__fact_payment_pos".event_dt "fact_payment_pos__event_dt",
              "fact_level_up_pos_key__fact_level_up_pos".event_dt "fact_level_up_pos__event_dt",
              "fact_level_up_pos_key__fact_level_up_pos".user_id "fact_level_up_pos__user_id"
            FROM
              public.guy_cube_test_payments AS "fact_level_up_pos_key__fact_payment_pos"
              LEFT JOIN (
                select
                  *
                from
(
                    select
                      event_dt,
                      new_level_id,
                      user_id,
                      cast(event_ts as date) as event_ss,
                      platform_id_psk
                    from
                      public.guy_cube_test_level
                    where
                      event_dt >= current_date -10
                  ) o
              ) AS "fact_level_up_pos_key__fact_level_up_pos" ON "fact_level_up_pos_key__fact_payment_pos".user_id = "fact_level_up_pos_key__fact_level_up_pos".user_id
              and "fact_level_up_pos_key__fact_payment_pos".event_dt = "fact_level_up_pos_key__fact_level_up_pos".event_dt
          ) AS "keys"
          LEFT JOIN (
            select
              *
            from
(
                select
                  event_dt,
                  new_level_id,
                  user_id,
                  cast(event_ts as date) as event_ss,
                  platform_id_psk
                from
                  public.guy_cube_test_level
                where
                  event_dt >= current_date -10
              ) o
          ) AS "fact_level_up_pos_key__fact_level_up_pos" ON "keys"."fact_level_up_pos__event_dt" = "fact_level_up_pos_key__fact_level_up_pos".event_dt
          AND "keys"."fact_level_up_pos__user_id" = "fact_level_up_pos_key__fact_level_up_pos".user_id
        GROUP BY
          1,
          2
      ) as q_0

-  ) AS "fact_payment_pos__level_subquery" ON "fact_payment_pos__level_subquery"."fact_payment_pos__user_id" = "fact_payment_pos".user_id,
-  "fact_payment_pos__level_subquery"."fact_payment_pos__event_dt" = "fact_payment_pos".event_dt
WHERE
  (
    "fact_payment_pos".event_ts >= $ 1 :: timestamptz
    AND "fact_payment_pos".event_ts <= $ 2 :: timestamptz
  )
GROUP BY
  1,
  2
ORDER BY
  3 DESC
LIMIT
  200000000
```

**Expected behavior**
Expected the join will be on user_id and event_dt 

Query example: 
```diff
SELECT
  "fact_payment_pos".user_id "fact_payment_pos__user_id",
  date_trunc(
    'day',
    (
      "fact_payment_pos".event_ts :: timestamptz AT TIME ZONE 'UTC'
    )
  ) "fact_payment_pos__event_ts_day",
  sum("fact_payment_pos".m_net_amount) "fact_payment_pos__sum_g",
  max("fact_payment_pos___level") "fact_payment_pos__level_max"
FROM
  public.guy_cube_test_payments AS "fact_payment_pos"
  LEFT JOIN (
    SELECT
      q_0."fact_payment_pos__user_id",
      q_0."fact_payment_pos__event_dt",
      "fact_level_up_pos__max_level" "fact_payment_pos___level"
    FROM
      (
        SELECT
          "keys"."fact_payment_pos__user_id",
          "keys"."fact_payment_pos__event_dt",
          max(
            "fact_level_up_pos_key__fact_level_up_pos".new_level_id
          ) "fact_level_up_pos__max_level"
        FROM
          (
            SELECT
              DISTINCT "fact_level_up_pos_key__fact_payment_pos".user_id "fact_payment_pos__user_id",
              "fact_level_up_pos_key__fact_payment_pos".event_dt "fact_payment_pos__event_dt",
              "fact_level_up_pos_key__fact_level_up_pos".event_dt "fact_level_up_pos__event_dt",
              "fact_level_up_pos_key__fact_level_up_pos".user_id "fact_level_up_pos__user_id"
            FROM
              public.guy_cube_test_payments AS "fact_level_up_pos_key__fact_payment_pos"
              LEFT JOIN (
                select
                  *
                from
(
                    select
                      event_dt,
                      new_level_id,
                      user_id,
                      cast(event_ts as date) as event_ss,
                      platform_id_psk
                    from
                      public.guy_cube_test_level
                    where
                      event_dt >= current_date -10
                  ) o
              ) AS "fact_level_up_pos_key__fact_level_up_pos" ON "fact_level_up_pos_key__fact_payment_pos".user_id = "fact_level_up_pos_key__fact_level_up_pos".user_id
              and "fact_level_up_pos_key__fact_payment_pos".event_dt = "fact_level_up_pos_key__fact_level_up_pos".event_dt
          ) AS "keys"
          LEFT JOIN (
            select
              *
            from
(
                select
                  event_dt,
                  new_level_id,
                  user_id,
                  cast(event_ts as date) as event_ss,
                  platform_id_psk
                from
                  public.guy_cube_test_level
                where
                  event_dt >= current_date -10
              ) o
+   ) AS "fact_level_up_pos_key__fact_level_up_pos" ON "keys"."fact_level_up_pos__event_dt" = "fact_level_up_pos_key__fact_level_up_pos".event_dt
+    AND "keys"."fact_level_up_pos__user_id" = "fact_level_up_pos_key__fact_level_up_pos".user_id
        GROUP BY
          1,
          2
      ) as q_0
  ) AS "fact_payment_pos__level_subquery" ON "fact_payment_pos__level_subquery"."fact_payment_pos__user_id" = "fact_payment_pos".user_id AND
  "fact_payment_pos__level_subquery"."fact_payment_pos__event_dt" = "fact_payment_pos".event_dt
WHERE
  (
    "fact_payment_pos".event_ts >= $ 1 :: timestamptz
    AND "fact_payment_pos".event_ts <= $ 2 :: timestamptz
  )
GROUP BY
  1,
  2
ORDER BY
  3 DESC
LIMIT
  200000000
```
**Version:**
0.33.23 - now

Additional context
checked on Postgres and Vertica drivers 
*Vertica - is currently WIP (will submit PR later on)

**Comment by guyzilb** [comment]
We are currently on formalizing PR for that

**Comment by guyzilb** [comment]
Yes, we working on the solution

בתאריך יום ג׳, 20 ביוני 2023 ב-21:52 מאת github-actions[bot] <
***@***.***>:

**Comment by igorlukanin** [comment]
Hey everyone, the fix has landed in [0.34.61](https://github.com/cube-js/cube/releases/tag/v0.34.61). Please feel free to upgrade!
